File: android/app/src/main/java/com/sand/table/Protocol/Serialization/Reader.java
package com.sand.table.Protocol.Serialization;

import com.sand.table.Model.Color;

import java.util.Arrays;

public class Reader {
    private final byte[] _data;
    private int _index = 0;

    public Reader(byte[] data) {
        _data = data;
    }

    public byte readByte() {
        return _data[_index++];
    }

    public byte[] readBytes() {
        return readBytes(_data.length - _index);
    }

    public byte[] readBytes(int length) {
        byte[] result = Arrays.copyOfRange(_data, _index, length);
        _index += length;
        return result;
    }

    public Color readColor() {
        return new Color(readByte(), readByte(), readByte());
    }

    public boolean readBoolean() {
        return readByte() != 0;
    }

    public short readShort() {
        int low = readByte() & 0xff;
        int high = readByte() & 0xff;
        return (short)(high << 8 | low);
    }

    public String readString() {
        StringBuilder result = new StringBuilder();
        while (true) {
            byte b = readByte();
            if (b == 0) {
                return result.toString();
            }

            result.append((char) b);
        }
    }
}

File: windows/Messages/Serialization/Reader.cs
using Sand.Model;
using Windows.Storage.Streams;

namespace Sand.Protocol.Serialization;

internal class Reader
{
    private DataReader _reader;

    public Reader(IBuffer data)
    {
        _reader = DataReader.FromBuffer(data);
        _reader.ByteOrder = ByteOrder.LittleEndian;
    }

    public void Read(out bool value) => value = _reader.ReadBoolean();

    public void Read(out byte value) => value = _reader.ReadByte();

    public void Read(out ushort value) => value = _reader.ReadUInt16();

    public void Read(out Color value)
    {
        value.r = _reader.ReadByte();
        value.g = _reader.ReadByte();
        value.b = _reader.ReadByte();
    }

    public void Read(out string value)
    {
        value = "";
        while (true)
        {
            var b = _reader.ReadByte();
            if (b == 0)
            {
                return;
            }

            value += (char)b;
        }
    }

    public IBuffer DetachBuffer() => _reader.DetachBuffer();
}
